Support backwards compatibility for v1beta1 custom metrics client

damemi opened the pull request:
Fixes #68011. The custom metrics client was updated to expect v1beta2 which broke all existing adapters. This reverts that change and adds a conversion from v1beta1 to v1beta2 internally

**What this PR does / why we need it**:
Fixes a bug in #64097 which broke custom_metrics clients for adapters using v1beta1

**Which issue(s) this PR fixes**
Fixes #68011

**Special notes for your reviewer**:

**Release note**:
```
NONE
```

damemi committed:
Support multiple versions in custom metrics client

This changes the custom metrics client logic over to support multiple versions
of the custom metrics API by checking discovery to find the appropriate versions.

Fixes #68011

Co-authored-by: Solly Ross <sross@redhat.com>

damemi commented:
/assign @DirectXMan12

DirectXMan12 commented:
@kubernetes/sig-autoscaling-pr-reviews

DirectXMan12 commented:
/milestone 1.12

DirectXMan12 commented:
/milestone v1.12

MaciekPytel commented on staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go:
```diff
@@ -237,13 +275,13 @@ func (m *namespacedMetrics) GetForObjects(groupKind schema.GroupKind, selector l
 	}
 
-	res := &v1beta2.MetricValueList{}
+	res := &v1beta1.MetricValueList{}
```
I don't really understand REST client machinery well enough, so I may very well be wrong on this, but won't this cause the following call to fail for an adapter that actually implements v1beta2? 

We initially put the result into v1beta1.MetricValueList{} and only later apply the conversion. But I don't expect v1beta2 result to be able to deserialize into a v1beta1 value list as MetricValue type is not compatible between versions. Am I missing something here?

DirectXMan12 replied:
Adapters will have to implement both APIs for the time being, and then we'll have to do the switch over in a later release.

So adapters should start implementing v1beta2 now, then when we switch over next release (or the one after) here, they'll have support for it, and a given adapter will work with both versions of kube.  Then, the release after that, we'll be able to finally remove v1beta1.

(that's the only way I can see this working, without fallback logic, and I don't think there's a precedence for fallback logic since we stopped using unversioned clients).

MaciekPytel replied:
I see your point. I wonder if we could address this by having a separate client for each version, discover which version is available and use the appropriate client in HPA?

I don't like the approach of switching the version at some point, mainly because I'm skeptical about every adapter implementing v1beta2 while it's not actually supported by k8s. I think we're effectively just pushing the problem to next release.

DirectXMan12 replied:
> I see your point. I wonder if we could address this by having a separate client for each version, discover which version is available and use the appropriate client in HPA?

I considered that, but it's way more complicated, and doesn't really seem to have a precendent in the rest of the codebase, IIRC (ever since we removed the unversioned clients).

> In the meantime every new adapter would have to implement a deprecated API.

Nobody needs to implement anything new except for the developers of kubernetes-incubator/custom-metrics-apiserver -- all adapters should get the APIs for free when they update.

Perhaps we should have @kubernetes/sig-api-machinery-pr-reviews weigh in?

deads2k replied:
@lavalamp another concrete example of the utility of a hub an a hubbed client.

@DirectXMan12 if it were my code, I would generate two different clients.  Then given an apiserver, I would use discovery to choose from the available APIs to new up the "best" client.  Then, if necessary, I would provide adapter clients that could skin one version as another one.  It's pretty painful, but with only two versions, it isn't so bad.

DirectXMan12 replied:
@deads2k ack, that seems reasonable to me eventually.  do we want to try to do it less than a week before code freeze while the tests are broken, though?

@damemi your mission, should you choose to accept it ^

DirectXMan12 commented:
/ok-to-test

DirectXMan12 commented:
/retest

DirectXMan12 commented:
/hold while we figure out what's going on

DirectXMan12 commented:
/hold

fedebongio commented:
/cc @wenjiaswe

damemi commented:
Cleaned up the fix and tested manually using a custom metrics adapter that expects v1beta1. Seems to work, just need to sort out the logic for determining the actual preferred version (instead of hardcoding v1beta1, which it is right now) but that's pretty much all there. Will update that and report back when this is ready for a final look before merge

DirectXMan12 commented:
SIG release: in case you're looking at this and want an update, and you can't reach me on Slack:

We either: a) need to merge the earlier version of this (relatively low risk)
 or b) need a bit more time (probably 1 or two days, depending on review latency) to finish and merge the fix suggested by @deads2k.  In either case, we need to merge *a* fix, as without it backwards compat with older CM API adapters is broken currently.

damemi commented:
@DirectXMan12 sorry forgot to comment but I did finish pushing the changes David recommended the other day, and the tests look green right now (also tested locally). 

I just don't like the approach of having to manually loop through all of the APIGroups and versions to find one that matches, but the Discovery client doesn't really provide a much better way unless I'm missing it. We could add a new helper to Discovery that maybe only returns the requested APIGroup but that's really just shifting the work around. If it's not worth it, then this fix is ready from my end

DirectXMan12 commented:
I'll need to do a quick review.  I'll try to get to it today, but chances are we'll need a short exception

liggitt commented:
> chances are we'll need a short exception

no exception needed for a bug fix restoring backwards compatibility prior to release, though a heads up to the release team would be appreciated, I'm sure.

Tagging the issue and/or into the milestone should get it on their radar as well.

DirectXMan12 commented on staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go:
```diff
@@ -20,27 +20,52 @@ import (
 }
 
-func New(client rest.Interface) CustomMetricsClient {
+func New(client rest.Interface, discoveryClient APIVersionsFromDiscovery) CustomMetricsClient {
```
this should be taking a `AvailableMetricsAPIFunc`, not an `APIVersionsFromDiscovery`.

DirectXMan12 replied:
ditto below

DirectXMan12 commented on staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go:
```diff
@@ -20,27 +20,52 @@ import (
+}
+
+func (d APIVersionsFromDiscovery) Versions() (*metav1.APIGroupList, error) {
+	groups, err := d.DiscoveryClient.ServerGroups()
```
this actual implementation should be filtering the groups by the metrics API -- it shouldn't be down the code in the metrics client.  This allows us to test the two independently.

DirectXMan12 replied:
(i.e. I'd lift the "find the metrics API group" out of the converter).

DirectXMan12 commented on cmd/kube-controller-manager/app/autoscaling.go:
```diff
@@ -48,9 +50,10 @@ func startHPAController(ctx ControllerContext) (http.Handler, bool, error) {
 	metricsClient := metrics.NewRESTMetricsClient(
 		resourceclient.NewForConfigOrDie(clientConfig),
-		custom_metrics.NewForConfigOrDie(clientConfig),
+		custom_metrics.NewForConfigOrDie(clientConfig, custom_metrics.APIVersionsFromDiscovery{DiscoveryClient: *discoveryClient}),
```
this is going to be checking discovery for *every* metrics request.  We should be caching the information and periodically rechecking it, or rechecking if we don't have any results.

DirectXMan12 reviewed (changes requested):
needs a few tweaks -- comments inline

DirectXMan12 committed:
Fix up potentially empty fields in HPA v2beta2

There were a few cases where fields that should have been marked optional
were not.  This was causing weird validation issues, but is now fixed.

DirectXMan12 commented:
/kind bug
/priority critical-urgent

(this is a non-trivial regression in compatibility)

DirectXMan12 commented:
/remove-sig architecture
/remove-sig networking
/remove-sig 
/remove-area kubelet 
/remove-sig node
/remove-sig testing

the autolabeller seems to have guessed wrong...

DirectXMan12 commented:
/remove-sig network
/remove-area kubelet

deads2k commented:
controller changes look ok.  You'll need an api-approver to put the API back.

/approve

smarterclayton commented:
/approve

API changes to optional

DirectXMan12 commented:
/lgtm

damemi commented:
/retest

damemi commented:
/test pull-kubernetes-e2e-kops-aws

damemi commented:
/test pull-kubernetes-e2e-kops-aws

nikhita commented:
@luxas looks like this should contain a release note?

DirectXMan12 commented:
@nikhita no, it shouldn't since it's a fix for an unintentional issue introduced in a previous PR this release.  This now works as noted in the original PR's release note.

nikhita commented:
@DirectXMan12 gotcha, thanks for the update!
